src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1453,6 +1523,31 @@ DatabaseShardImp::storeLedgerInShard(
     return result;
 }
 
+void
+DatabaseShardImp::removeFailedShard(
+    std::shared_ptr<Shard> shard,
+    std::lock_guard<std::mutex> & lock)
+{
+    shards_.erase(shard->index());
+    if (shard->index() == acquireIndex_)
+        acquireIndex_ = 0;
+
+    updateStatus(lock);
+
+    using namespace boost::filesystem;
+    path const dir {shard->getDir()};
+    shard.reset();
```

Comment:
We are holding the lock unnecessarily while destroying the shard and removing directory. That could be a problem if the OS file system is busy.

Reply:
When I replaced the duplicated code with this function I made sure the lock was already being held in those spots, so I don't think this function will cause any new issues.

Reply:
Narrowed the lock scope per our conversation.